MSC4335: M_USER_LIMIT_EXCEEDED error code (#4335)
matrix-org/matrix-doc: proposals/4335-user-limit-exceeded.md

resource constraints, and overloading it could create confusion about whether the limit applies
to the user specifically or the server generally.

**Add structured error information**: Instead of a simple error code, a more complex error format
could include machine-readable fields for limit types, current usage, and maximum limits. While
this would provide more information, it would require a more significant change to the error
response format and could be added in a future MSC if needed.

**Multiple specific error codes**: Separate error codes could be introduced for different types
of limits (e.g., `M_STORAGE_LIMIT_EXCEEDED`, `M_ROOM_LIMIT_EXCEEDED`). However, this approach
would require many new error codes and doesn't provide significant benefits over a single code
with descriptive error messages.

## Security considerations

None as only adding a new error code.

## Unstable prefix

While this proposal is being developed and refined, implementations should use the prefix the
`errcode` and `info_url` fields with `org.matrix.msc4335.` and use a value of `M_UNKNOWN` for
`errcode`.

For example:

```json
{
  "errcode": "M_UNKNOWN",

>   "org.matrix.msc4335.errcode": "M_USER_LIMIT_EXCEEDED",

Reviewer: No need to change anything since it appears to already be implemented like this but I was slightly surprised that the prefix is on the key rather than the value. Would this not hypothetically allow for odd situations where you have `errcode`s from several MSCs, possibly leaving clients confused about which one to choose? I couldn't find any guidance on where to put the prefix in https://spec.matrix.org/proposals/ FWIW.

Author: It would be great to get some clarification on what is the "right" way to do this.

Re-reading https://spec.matrix.org/v1.16/client-server-api/#standard-error-response I see it does state how custom error codes should be written.

I think this implicitly means that clients *should* expect non-standard error codes and do something sensible with them (which was what I was assuming would not be the case and hence prefixing the key).

* `POST /_matrix/media/v3/upload` - When storage quota is exceeded
* `POST /_matrix/client/v3/rooms/{roomId}/invite` - When invite limits (like maximum participant count) are exceeded

The HTTP response code should be chosen based on the specification for the individual endpoint. For
example, the most appropriate code for `POST /_matrix/media/v3/upload` would be `403`.

### Distinction from Other Error Codes

This error code is distinct from:

* `M_LIMIT_EXCEEDED`: Used for general rate limiting that applies to all users or based on IP/client
* `M_FORBIDDEN`: Used for authorization failures or policy violations not related to usage limits
* `M_RESOURCE_LIMIT_EXCEEDED`: Used for server-wide resource constraints affecting all users
* `M_TOO_LARGE`: Used when a request is too large (file size, message length, etc.) regardless of user limits

## Potential issues

This error code does not specify the exact nature of the limit that was exceeded, which could
potentially lead to ambiguity. However, this is consistent with other Matrix error codes that
rely on the human-readable `error` field to provide specific details. Implementers should ensure

> they provide clear, actionable error messages.

Author: I've updated the proposal in [`fbd628a` (#4335)](https://github.com/matrix-org/matrix-spec-proposals/pull/4335/commits/fbd628a5cd39e7e23973a226c8348a7098c42f23) to include an `info_link` field.

> they provide clear, actionable error messages.

Reviewer 2: Your commit says `info_link` but the markdown says `info_url` :D

might be enforced. Examples could include:

* [`POST /_matrix/media/v3/upload`] - When storage quota is exceeded
* [`POST /_matrix/client/v3/rooms/{roomId}/invite`] - When invite limits (like maximum participant count) are exceeded

### Distinction from Other Error Codes

This error code is distinct from:

* `M_LIMIT_EXCEEDED`: Used for general rate limiting that applies to all users or based on IP/client
* `M_FORBIDDEN`: Used for authorization failures or policy violations not related to usage limits
* `M_RESOURCE_LIMIT_EXCEEDED`: Used for server-wide resource constraints affecting all users
* `M_TOO_LARGE`: Used when a request is too large (file size, message length, etc.) regardless of user limits

## Potential issues

This error code does not specify the exact nature of the limit that was exceeded, which could
potentially lead to ambiguity. However, this is consistent with other Matrix error codes that
rely on the human-readable `error` field to provide specific details. Instead the `info_uri`
provides a way for the homeserver to apply arbitrary limits without the client having to understand

> every type in advance.

Author: Screenshots are at the top of this PR. https://github.com/matrix-org/matrix-spec-proposals/pull/4335#issue-3367446512

> every type in advance.

Author: In the context of Synapse and the media upload limits feature, I imagined it could be configured with context like this:

```yaml
media_upload_limits:
- time_period: 24h
  max_size: 100M
  msc4335_info_uri: https://example.com/quota?type=daily
  msc4335_soft_limit: true
  msc4335_increase_uri: https://example.com/increase-quota?type=daily
- time_period: 28d
  max_size: 500M
  msc4335_info_uri: https://example.com/quota?type=monthly
  msc4335_soft_limit: true
  msc4335_increase_uri: https://example.com/increase-quota?type=monthly
```

Which would result in error response such as:

```json
{
  "errcode": "M_USER_LIMIT_EXCEEDED",
  "error": "User has exceeded their upload limit",
  "info_uri": "https://example.com/quota?type=daily",
  "soft_limit": true,
  "increase_uri": "https://example.com/increase-quota?type=daily"
}
```

The web page at the other end can then serve up content that is specific to the limit exceeded.

Generalising this, the homeserver implementation and/or operator could also do things like:

- Make use of path: https://example.com/quota-monthly
- Fragment: https://example.com/quota#monthly
- Multiple params: https://example.com/plan?action=limit_exceeded&type=upload_limit&period=monthly

In order to implement the soft limit and option to upgrade, a Synapse operator would need to make use the [get_media_upload_limits_for_user](https://element-hq.github.io/synapse/latest/modules/media_repository_callbacks.html#get_media_upload_limits_for_user) from the module API to allow certain users to have different media upload limits.

@clokep is it that you prefer the approach described at https://github.com/matrix-org/matrix-spec-proposals/blob/hughns/user-limit-exceeded/proposals/4335-user-limit-exceeded.md#add-structured-error-information? If so then could you add a "vote" or preference to that alternative instead?

If not, then I feel like I am missing something.

> every type in advance.

Reviewer 3: OK, that sounds fine. I think that's not clear from the text that a _different_ URI per resource is expected. And the matching implementation doesn't do this.

The homeserver can choose to provide localised and personalised content on the `info_uri`, `soft_limit` and
`increase_uri` if it wishes.

The error code does not provide machine-readable information about current usage or limits,
which could be useful for clients to display progress bars or usage statistics. However, adding
such fields would require a more complex specification change and could be addressed in a future
MSC if deemed necessary.

## Alternatives

Several alternatives were considered for this proposal:

**Use M_RESOURCE_LIMIT_EXCEEDED**: The existing [`M_RESOURCE_LIMIT_EXCEEDED`] error code could be
expanded to cover user-specific limits. However, this code is currently used for server-wide
resource constraints, and overloading it could create confusion about whether the limit applies

> to the user specifically or the server generally.

Reviewer: All excellent. Thank you. 👍

limits that are specifically associated with their account, such as:

* **Storage quotas**: When a user has exceeded their allocated storage space for media uploads,
  message history, or other persistent data.
* **Resource limits**: When a user has reached their maximum number of allowed rooms, devices,
  or other account-scoped resources.
* **Feature limits**: When a user has exceeded usage limits for specific features (e.g., number
  of public rooms they can create, number of invites they can send).
* **Account tier restrictions**: When a user's account type (free, premium, etc.) prevents them
  from performing certain operations.

The error response must also contain additional fields:

* `info_uri` string (required) - an opaque URI that the client can link the user to in order to get more context on the
  encountered limit.
* `soft_limit` boolean (optional, default `false`) - `true` means that the specific limit encountered can be increased.
  Otherwise it is a hard limit that cannot be increased.
* `increase_uri` (required if `soft_limit` is `true`) - an opaque URI where the user can undertake actions to increase
  the encountered limit.

> The `info_uri` and `increase_uri` are "opaque" in the sense that the homeserver implementation may choose to encode

Author: I agree that this is unclear. In 737f53bfac518a920df44cd05c3c560e7db507a0 I have removed `increase_uri` entirely and also renamed the `soft_limit` field to `can_upgrade`.

information, such as the type of limit encountered, within the URI but it may do so using an encoding of its choosing.

The homeserver may return *different* values for `info_uri` and `increase_uri` depending on what type of limit was
reached.

The HTTP response code should be chosen based on the specification for the individual endpoint. For
example, the most appropriate code for [`POST /_matrix/media/v3/upload`] would be `403 Forbidden`.

An example response body for the error might look as follows:

```json
{
  "errcode": "M_USER_LIMIT_EXCEEDED",

>   "error": "User has exceeded their storage quota of 10GB",

Reviewer 4: nit: since clients sometimes show the `error` to the user, I think

```suggestion
  "error": "You have exceeded your storage quota of 10GB",
```

would look better. (same below)

    "fr": "Vous avez atteint votre limite de téléchargement quotidienne. Pour plus d'informations sur les limites d'utilisation applicables à votre compte, cliquez <a href=\"https://matrix.org/homeserver/pricing/#usage-limits\">ici</a>. Vous pouvez passer à un compte Premium pour augmenter ces limites <a href=\"https://account.matrix.org/account?action=org.matrix.plan_management\">ici</a>.",
  }
}
```

Similarly for a "hard" limit:

```json
{
  "errcode": "M_USER_LIMIT_EXCEEDED",
  "error": "User has exceeded a media upload limit",
  "messages": {
    "en": "You have reached your daily upload limit. More information on the usage limits that apply to your account is available <a href=\"https://matrix.org/homeserver/pricing/#usage-limits\">here</a>.",
    "fr": "Vous avez atteint votre limite de téléchargement quotidienne. Pour plus d'informations sur les limites d'utilisation applicables à votre compte, cliquez <a href=\"https://matrix.org/homeserver/pricing/#usage-limits\">ici</a>.",
  }
}
```

Comparison:

* An advantage is that the homeserver can be as specific as it wishes with the messages. For example, the matrix.org

>   homeserver can specifically refer to the usage plans that are available rather than having to be generic.

Reviewer 2: Clients shouldn't show a non-i18n'd error field though because it may not match the user's language requirements, so clients end up having to use only `errcode` and mapping those to translations themselves

>   homeserver can specifically refer to the usage plans that are available rather than having to be generic.

Reviewer 5: (considering resolved - please open a new thread if untrue)

  "errcode": "ORG.MATRIX.MSC4335_USER_LIMIT_EXCEEDED",
  "error": "User has exceeded their fair usage limit of 2GB",
  "org.matrix.msc4335.info_uri": "https://example.com/homeserver/about?limit_type=quota",
  "org.matrix.msc4335.can_upgrade": true
}
```

### Include a URI to complete the upgrade action

An earlier version of the proposal included an `increase_uri` that could be attached as a UI button for the user to
take start the action of upgrading.

However, this led to [uncertainty](https://github.com/matrix-org/matrix-spec-proposals/pull/4335#discussion_r2582153412)
about what exactly would happen when the button is pressed.

> To remove this uncertainty `increase_uri` was removed and the homserver gets to provide a single `info_uri` and it is

Reviewer 6: ```suggestion
To remove this uncertainty `increase_uri` was removed and the homeserver gets to provide a single `info_uri` and it is
```
